### PR TITLE
update GAN

### PR DESCRIPTION
**Pull request by MorganeLu**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
- [ ] review the artifacts from local build
- [ ] [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [ ] Unit test if appliable

# Checklist:

- [ ] My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have applied tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 1]
```diff
@@ -13,5 +13,274 @@ kernelspec:
...
+
+### Generator gradient for update (gradient descent)
+
+The aim of Generator is $\bigtriangledown_{W_G} \frac{1}{n} \sum_{i=1}^n log(1-D(G(z^{(i)})))$. 
+$D(G(z^{(i)}))$ is also used to predict fake images, but if it predicts badly on fake images, the probability will be close to 1, which is not same to discriminator.
+
+### Algorithm
+
+:::{figure-md} GAN algorithm
+<img src="../../images/deep-learning/GAN/algorithm.png" width="90%" class="bg-white mb-1">
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/GAN.md:61]: We aim to be math-light, so such algorithms should not be included.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 2]
```diff
@@ -13,5 +13,274 @@ kernelspec:
...
+
+from utils import Logger
+
+import tensorflow as tf
+import tensorflow.compat.v1 as tf
+tf.disable_v2_behavior() 
+
+import numpy as np
+
+DATA_FOLDER = './tf_data/VGAN/MNIST'
```
Comment (reveurmichael) [inline open-machine-learning-jupyter-book/deep-learning/GAN.md:104]: We don't have such dataset, right?

Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:104]: yes, I did not find the proper codes for GANs, and this dataset is the normal MNIST dataset and it is open-source and available.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 1 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 3]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+    return tf.random_uniform(shape=size, minval=-stddev, maxval=stddev)
+
+def images_to_vectors(images):
+    return images.reshape(images.shape[0], 784)
+
+def vectors_to_images(vectors):
+    return vectors.reshape(vectors.shape[0], 28, 28, 1)
+```
+
+Load Data
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/GAN.md:117]: Is this a title? If it is, make it as a title.

Also, give an empty line before the `{code-cell}`

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 4]
```diff
@@ -13,5 +13,267 @@ kernelspec:
   name: python3
 ---
 
-# GAN (TBD)
+# Generative adversarial networks 
 
+The original purpose of GAN is to generate new data. It classically generates new images, but is applicable to wide range of domains.
+It learns the training set distribution and can generate new images that have never been seen before.
+In contrast to e.g., autoregressive models or RNNs (generating one word at a time), GANs generate the whole output all at once.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/GAN.md:20]: We must be consistent for the name. We will use GANs or GAN?

[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 5]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
 
-# GAN (TBD)
+# Generative adversarial networks 
 
+The original purpose of GAN is to generate new data. It classically generates new images, but is applicable to wide range of domains.
+It learns the training set distribution and can generate new images that have never been seen before.
+In contrast to e.g., autoregressive models or RNNs (generating one word at a time), GANs generate the whole output all at once.
+
+```{seealso}
+GAN is proposed in 2005, the paper is [Generative Adversarial Networks](https://arxiv.org/abs/1406.2661).
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Yes, I chose the MLA style but I am not sure whether I wrote in the right way. If possible, could you help me to revise it?

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 6]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+```
+
+## overview
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 7]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+
+The structure of GAN is roughly as follows. It has two essential part, discriminator and generator. 
+Discriminator learns to become better at distinguishing real from generated images, and generator learns to generate better images to fool the discriminator.
+
+:::{figure-md} GAN structure
+<img src="../../images/deep-learning/GAN/structure.png" width="90%" class="bg-white mb-1">
+
+Illustration of GAN structure
+:::
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:33]: This problem has been discussed before. Since the images are all form the projects I mentioned in the acknowledge part, so another reference is not uncessary.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 8]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+
+Then, one question appears: why Are GANs Are Called Generative Models?
+-The generative part comes from the fact that the model "generates" new data,
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 9]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+The structure of GAN is roughly as follows. It has two essential part, discriminator and generator. 
+Discriminator learns to become better at distinguishing real from generated images, and generator learns to generate better images to fool the discriminator.
+
+:::{figure-md} GAN structure
+<img src="../../images/deep-learning/GAN/structure.png" width="90%" class="bg-white mb-1">
+
+Illustration of GAN structure
+:::
+
+Then, one question appears: why Are GANs Are Called Generative Models?
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done, sorry for my mistakes.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 10]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+-The generative part comes from the fact that the model "generates" new data,
+- Usually, generative models use an approximation to compute the usually intractable distribution; here, the discriminator part does that approximation,
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: I have replaced the first usually with "most of the time".

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 11]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+## Training
+
+To make the model converge, we need to find a proper objective for it.
+For the GAN, it is $\min\limits_{G} \max\limits_{D} V(D, G) = \mathbb{E}_{x \thicksim p_{data}}[log D(x)] + \mathbb{E}_{z\thicksim p_z(z)}[log(1-D(G(z)))]$
+
+### Discriminator gradient for update (gradient ascent)
+
+The aim of Discriminator is $\bigtriangledown_{W_D} \frac{1}{n} \sum_{i=1}^n [log D(x^{(i)}) + log (1-D(G(z^{(i)})))]$. We can split it into two parts.
+First is $D(x^{(i)})$. If it predicts well in real images, the probability will be close to 1. 
+Second is $D(G(z^{(i)}))$. This part use to predict the fake images which generdated from the generator, and if it predicts well, the probability will be close to 0.
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 12]
```diff
@@ -13,5 +13,267 @@ kernelspec:
...
+
+### Generator gradient for update (gradient descent)
+
+The aim of Generator is $\bigtriangledown_{W_G} \frac{1}{n} \sum_{i=1}^n log(1-D(G(z^{(i)})))$. 
+$D(G(z^{(i)}))$ is also used to predict fake images, but if it predicts badly on fake images, the probability will be close to 1, which is not same to discriminator.
+
+### Train for convergence
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:56]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 3 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:118]: I don't regard it as a title, and I think it just tell us the aim of the following code cell.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 4 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 2 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:100]: I have added the download command into the code cell.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 13]
```diff
@@ -13,5 +13,267 @@ kernelspec:
   name: python3
 ---
 
-# GAN (TBD)
+# Generative adversarial networks
```
[the remark this answers is not in the document]
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:16]: Yes, but after I changed the file name into Generative adversarial networks, the GAN file is still there. Or you would like to keep the title as GAN?

File: open-machine-learning-jupyter-book/deep-learning/Generative-adversarial-networks.md [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.11.5
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+# Generative adversarial networks
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/Generative-adversarial-networks.md:16]: The file name must be all lowercase.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:16]: You can delete it in a separate PR when you add the file to the toc.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md [hunk 15]
```diff
@@ -13,5 +13,275 @@ kernelspec:
   name: python3
 ---
 
-# GAN (TBD)
+# Generative adversarial networks 
 
+The original purpose of Generative adversarial networks(GAN) is to generate new data. It classically generates new images, but is applicable to wide range of domains. It learns the training set distribution and can generate new images that have never been seen before. In contrast to e.g., autoregressive models or RNNs (generating one word at a time), GANs generate the whole output all at once.
+
+```{seealso}
+GAN is proposed in 2005, the paper is Ian J. Goodfellow, Jean Pouget-Abadie, Mehdi Mirza, Bing Xu, David Warde-Farley, Sherjil Ozair, Aaron Courville, Yoshua Bengio. "Generative Adversarial Networks", arxiv:1406.2661.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/GAN.md:21]: Just put the reference is enough. Don't need to mention 'GAN is proposed'.

File: open-machine-learning-jupyter-book/deep-learning/GAN.md (hunk 2 above)
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/deep-learning/GAN.md:100]: Please make sure this file could pass the build.

**Review by zhangqi444 — APPROVED** [review]
